src/core/service.c

```diff
@@ -1646,15 +1646,10 @@ static Service *service_get_triggering_service(Service *s) {
          *
          * N.B. if there are multiple services which could trigger 's' via OnFailure=
          * or OnSuccess= then we return NULL. This is since we don't know from which
-         * one to propagate the exit status. */
+         * one to propagate the exit status. Only exception is if the same service
+         * lists this service as their OnFailure= and OnSuccess= handler. */
 
-        UNIT_FOREACH_DEPENDENCY(other, UNIT(s), UNIT_ATOM_ON_FAILURE_OF) {
-                if (candidate)
-                        goto have_other;
-                candidate = other;
-        }
-
-        UNIT_FOREACH_DEPENDENCY(other, UNIT(s), UNIT_ATOM_ON_SUCCESS_OF) {
+        UNIT_FOREACH_DEPENDENCY(other, UNIT(s), UNIT_ATOM_ON_FAILURE_OF|UNIT_ATOM_ON_SUCCESS_OF) {
```

Comment:
The comments say that `UNIT_FOREACH_DEPENDENCY` checks for any of the atoms and not all. This does not seem to be used anywhere in the code? Is it discouraged?

Reply:
`unit_success_failure_handler_has_jobs` similarly has two loops with the respective atoms instead of a single one with the atom flags merged. The comments state that there is some optimization if the atom is unique?

If having this as two separate loop is better for performance I can split them again and add a `if (candidate && other != candidate)` instead... (or whatever one uses to compare units for equality)